pkgs/build-support/mkshell/default.nix

```diff
@@ -28,18 +28,63 @@ let
     "propagatedNativeBuildInputs"
     "shellHook"
   ];
+
+  merged = {
+    buildInputs = mergeInputs "buildInputs";
+    nativeBuildInputs = packages ++ (mergeInputs "nativeBuildInputs");
+    propagatedBuildInputs = mergeInputs "propagatedBuildInputs";
+    propagatedNativeBuildInputs = mergeInputs "propagatedNativeBuildInputs";
+  };
+
+  docPathShellHook = let
+    unwrapCC = pkg: if pkg ? cc then unwrapCC pkg.cc else pkg;
+    docInputs = builtins.concatLists (builtins.attrValues merged)
+                # grab the unwrapped cc, as wrappers may or may not
+                # retain doc outputs
+                ++ (lib.optional (stdenv ? cc) (unwrapCC stdenv.cc))
+                ++ stdenv.defaultBuildInputs ++ stdenv.defaultNativeBuildInputs
+                ++ stdenv.initialPath;
+    docDrvs = lib.unique (builtins.filter lib.isDerivation docInputs);
```

Comment:
`lib.unique` has O(n²) complexity, can this be moved to build time or dropped?

Reply:
Simplified, _all_ nix code is evaluated before any build phases is run _(unless you do "import from derivation" which is banned in nixpkgs)_, the result of which is stored in a `.drv` file in the store. Once the `.drv` file is written all relevant nix code must have been evaluated. 

That `.drv` file essentially contains a bunch of build stages as strings, and is _realized_ by `nix-build` either by pulling from the online cache or by performing the build locally. During realization/build the different build stage are executed. `nix-shell` will perform a _partial_ build and stop right before `buildPhase`, and then run `shellHook`. The `shellHook` string is however evaluated before the build, since the shell you build and enter is a derivation.